src/test/java/org/jenkinsci/plugins/workflow/job/properties/DisableConcurrentBuildsJobPropertyTest.java

```diff
@@ -79,5 +86,62 @@ public void concurrentBuildsMigrationOnByDefault() throws Exception {
 
         WorkflowJob roundTripDisabled = r.configRoundtrip(disabledCase);
         assertFalse(roundTripDisabled.isConcurrentBuild());
+
+        DisableConcurrentBuildsJobProperty prop = roundTripDisabled.getProperty(DisableConcurrentBuildsJobProperty.class);
+        assertNotNull(prop);
+        assertFalse(prop.isAbortPrevious());
+        prop.setAbortPrevious(true);
+        roundTripDisabled = r.configRoundtrip(roundTripDisabled);
+        assertTrue(roundTripDisabled.isConcurrentBuild()); // see comment there
+        prop = roundTripDisabled.getProperty(DisableConcurrentBuildsJobProperty.class);
+        assertNotNull(prop);
+        assertTrue(prop.isAbortPrevious());
     }
+
+    @Issue("JENKINS-43353")
+    @Test public void abortPrevious() throws Exception {
+        WorkflowJob p = r.createProject(WorkflowJob.class);
+        p.setDefinition(new CpsFlowDefinition("semaphore 'run'", true));
+
+        // Control case: concurrent builds allowed.
+        assertTrue(p.isConcurrentBuild());
+        WorkflowRun b1 = p.scheduleBuild2(0).waitForStart();
+        SemaphoreStep.waitForStart("run/1", b1);
+        WorkflowRun b2 = p.scheduleBuild2(0).waitForStart();
+        SemaphoreStep.waitForStart("run/2", b2);
+        SemaphoreStep.success("run/1", null);
+        SemaphoreStep.success("run/2", null);
+        r.waitForCompletion(b1);
+        r.waitForCompletion(b2);
+
+        // Control case: simple disable concurrent.
+        p.setConcurrentBuild(false);
+        assertFalse(p.isConcurrentBuild());
+        WorkflowRun b3 = p.scheduleBuild2(0).waitForStart();
+        SemaphoreStep.waitForStart("run/3", b3);
+        SemaphoreStep.success("run/4", null);
+        QueueTaskFuture<WorkflowRun> b4f = p.scheduleBuild2(0);
+        Thread.sleep(1000); // TODO is there a cleaner way for the queue to finish processing?
```

Comment:
Perhaps `Queue#maintain` as in https://github.com/jenkinsci/throttle-concurrent-builds-plugin/blob/05f713e9a3c41674b0a963c77e8b2a51e5be6206/src/test/java/hudson/plugins/throttleconcurrents/ThrottleJobPropertyPipelineTest.java#L98 ?

Reply:
Afraid not.

```diff
diff --git src/test/java/org/jenkinsci/plugins/workflow/job/properties/DisableConcurrentBuildsJobPropertyTest.java src/test/java/org/jenkinsci/plugins/workflow/job/properties/DisableConcurrentBuildsJobPropertyTest.java
index 1c19a29..780586f 100644
--- src/test/java/org/jenkinsci/plugins/workflow/job/properties/DisableConcurrentBuildsJobPropertyTest.java
+++ src/test/java/org/jenkinsci/plugins/workflow/job/properties/DisableConcurrentBuildsJobPropertyTest.java
@@ -115,13 +115,11 @@ public class DisableConcurrentBuildsJobPropertyTest {
         r.waitForCompletion(b2);
 
         // Control case: simple disable concurrent.
-        p.setConcurrentBuild(false);
-        assertFalse(p.isConcurrentBuild());
         WorkflowRun b3 = p.scheduleBuild2(0).waitForStart();
         SemaphoreStep.waitForStart("run/3", b3);
         SemaphoreStep.success("run/4", null);
         QueueTaskFuture<WorkflowRun> b4f = p.scheduleBuild2(0);
-        Thread.sleep(1000); // TODO is there a cleaner way for the queue to finish processing?
+        r.jenkins.getQueue().maintain();
         assertFalse(b4f.getStartCondition().isDone());
         SemaphoreStep.success("run/3", null);
         r.waitForCompletion(b4f.waitForStart());
```

falls through to an NPE at https://github.com/jenkinsci/workflow-job-plugin/blob/1a07b9ff3ff4336e020a5076dd72e776ba4845e8/src/test/java/org/jenkinsci/plugins/workflow/job/properties/DisableConcurrentBuildsJobPropertyTest.java#L131 while

```diff
diff --git src/test/java/org/jenkinsci/plugins/workflow/job/properties/DisableConcurrentBuildsJobPropertyTest.java src/test/java/org/jenkinsci/plugins/workflow/job/properties/DisableConcurrentBuildsJobPropertyTest.java
index 1c19a29..7c06f85 100644
--- src/test/java/org/jenkinsci/plugins/workflow/job/properties/DisableConcurrentBuildsJobPropertyTest.java
+++ src/test/java/org/jenkinsci/plugins/workflow/job/properties/DisableConcurrentBuildsJobPropertyTest.java
@@ -115,12 +115,11 @@ public class DisableConcurrentBuildsJobPropertyTest {
         r.waitForCompletion(b2);
 
         // Control case: simple disable concurrent.
-        p.setConcurrentBuild(false);
-        assertFalse(p.isConcurrentBuild());
         WorkflowRun b3 = p.scheduleBuild2(0).waitForStart();
         SemaphoreStep.waitForStart("run/3", b3);
         SemaphoreStep.success("run/4", null);
         QueueTaskFuture<WorkflowRun> b4f = p.scheduleBuild2(0);
+        r.jenkins.getQueue().maintain();
         Thread.sleep(1000); // TODO is there a cleaner way for the queue to finish processing?
         assertFalse(b4f.getStartCondition().isDone());
         SemaphoreStep.success("run/3", null);
```

fails at https://github.com/jenkinsci/workflow-job-plugin/blob/1a07b9ff3ff4336e020a5076dd72e776ba4845e8/src/test/java/org/jenkinsci/plugins/workflow/job/properties/DisableConcurrentBuildsJobPropertyTest.java#L125

This test really needs to check that not only does `Queue` finish its work, but that there is not a new `WorkflowRun` started, which I think is a separate asynchronous process.

Reply:
Oh I see, checking `CauseOfBlockage` helps. Working on it…

Reply:
Indeed

```diff
diff --git src/test/java/org/jenkinsci/plugins/workflow/job/properties/DisableConcurrentBuildsJobPropertyTest.java src/test/java/org/jenkinsci/plugins/workflow/job/properties/DisableConcurrentBuildsJobPropertyTest.java
index 1c19a29..88bfa62 100644
--- src/test/java/org/jenkinsci/plugins/workflow/job/properties/DisableConcurrentBuildsJobPropertyTest.java
+++ src/test/java/org/jenkinsci/plugins/workflow/job/properties/DisableConcurrentBuildsJobPropertyTest.java
@@ -27,6 +27,10 @@ package org.jenkinsci.plugins.workflow.job.properties;
 
 import hudson.model.Result;
 import hudson.model.queue.QueueTaskFuture;
+import java.util.Collections;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import jenkins.model.BlockedBecauseOfBuildInProgress;
 import jenkins.model.InterruptedBuildAction;
 import org.jenkinsci.plugins.workflow.cps.CpsFlowDefinition;
 import org.jenkinsci.plugins.workflow.job.WorkflowJob;
@@ -115,14 +119,12 @@ public class DisableConcurrentBuildsJobPropertyTest {
         r.waitForCompletion(b2);
 
         // Control case: simple disable concurrent.
-        p.setConcurrentBuild(false);
-        assertFalse(p.isConcurrentBuild());
         WorkflowRun b3 = p.scheduleBuild2(0).waitForStart();
         SemaphoreStep.waitForStart("run/3", b3);
         SemaphoreStep.success("run/4", null);
         QueueTaskFuture<WorkflowRun> b4f = p.scheduleBuild2(0);
-        Thread.sleep(1000); // TODO is there a cleaner way for the queue to finish processing?
-        assertFalse(b4f.getStartCondition().isDone());
+        r.jenkins.getQueue().maintain();
+        assertEquals(Collections.singletonList(BlockedBecauseOfBuildInProgress.class), r.jenkins.getQueue().getItems(p).stream().map(i -> i.getCauseOfBlockage().getClass()).collect(Collectors.toList()));
         SemaphoreStep.success("run/3", null);
         r.waitForCompletion(b4f.waitForStart());
         r.waitForCompletion(b3);
```

fails as expected:

```
expected:<[class jenkins.model.BlockedBecauseOfBuildInProgress]> but was:<[class hudson.model.queue.CauseOfBlockage$NeedsMoreExecutorImpl]>
```

(Well, sort of. I would not have expected a flyweight task to have `NeedsMoreExecutorImpl`; would have expected an empty list or something. But close enough.)